In [189]:
import pandas as pd

In [190]:
# Load posts and reels data
df_posts = pd.read_csv('instagram/Instagram - Posts.csv')
df_reels = pd.read_csv('instagram/Instagram - Reels.csv')

### for the hashtags section, we will only need to consider the hashtags column as the independent variable, as well as the engagement metrics to measure virality as the dependent variable, namely num_comments and likes. We will consider views for reels separately since theres no views column for posts.

In [191]:
# Combine hashtags and engagement metrics
posts_subset = df_posts[['hashtags', 'num_comments', 'likes']]
reels_subset = df_reels[['hashtags', 'num_comments', 'likes', 'views', 'followers']]
df_combined = pd.concat([posts_subset, reels_subset], ignore_index=True)
df_combined

,hashtags,num_comments,likes,views,followers
0,"[""#พร้อมส่ง"",""#สายฝอ"",""#สายฝอพร้อมส่ง"",""#สายฝอ...",0,19.0,NaN,NaN
1,"[""#GanpatiBappaMorya"",""#GaneshChaturthi"",""#Fes...",14,189.0,NaN,NaN
2,"[""#ramadan"",""#ramadankareem"",""#ramadanmubarak""...",0,68.0,NaN,NaN
3,NaN,1,NaN,NaN,NaN
4,NaN,10,397.0,NaN,NaN
...,...,...,...,...,...
1995,NaN,4,493.0,27696.0,31712.0
1996,"[""#PINPrestigeMalaysia"",""#PINMYCover"",""#coverg...",0,-1.0,339.0,NaN
1997,"[""#turtle"",""#rehabilitation"",""#maldives"",""#ree...",1,188.0,2381.0,35319.0
1998,"[""#iOS"",""#AppleIntelligence"",""#Apple"",""#iPhone...",2,-1.0,1115.0,NaN


### We assume that NaN rows in hashtags means the post does not have any hashtags. We will now check for NaN or <0 rows for num_comments, likes, views, followers.

In [192]:
# Function to check for problematic rows
def checkForProblematicRows(df):
    df_check = df.copy()
    
    # For views, we will start checking from row 1000 onwards (ignoring first 1000 rows because they are definitely NaN)
    views = (df_check['views'].isna() | (df_check['views'] < 0)) & (df_check.index >= 1000)
    
    # For num_comments and likes, check all rows
    comments = df_check['num_comments'].isna() | (df_check['num_comments'] < 0)
    likes = df_check['likes'].isna() | (df_check['likes'] < 0)

    # For followers, we will start checking from row 1000 onwards as well.
    followers = (df_check['followers'].isna() | (df_check['followers'] < 0)) & (df_check.index >= 1000)
    
    # Combine all conditions
    final_combined_attributes = comments | likes | views | followers
    
    # Get the rows that meet any of the conditions
    problematic_rows = df_check[final_combined_attributes]

    return problematic_rows

In [193]:
problematic_rows = checkForProblematicRows(df_combined)
problematic_rows

,hashtags,num_comments,likes,views,followers
3,NaN,1,NaN,NaN,NaN
10,"[""#lookoftheday"",""#beyourownkindofbeautiful"",""...",22,NaN,NaN,NaN
11,"[""#myforeverlove""]",61,NaN,NaN,NaN
13,NaN,0,NaN,NaN,NaN
16,"[""#kanemotoyoshinori"",""#yoshi""]",0,NaN,NaN,NaN
...,...,...,...,...,...
1989,"[""#Repost""]",5,101.0,523.0,NaN
1991,NaN,0,50.0,815.0,NaN
1993,"[""#کبوتر_کفتر_عشق_دهاتی_تیکه_پاره_ساعت_سنگین_ت...",72,2827.0,39111.0,NaN
1996,"[""#PINPrestigeMalaysia"",""#PINMYCover"",""#coverg...",0,-1.0,339.0,NaN


### We will now eliminate the NaN values for each column.

### For NaN rows for hashtags, we will assume that the user did not include any hashtags, and so we will replace the NaN rows with empty lists instead.

### For NaN and negative rows for num_comments, likes,views, followers, we will set 0 values, assuming that NaN values indicate that there are no values respectively and negative values being impossible to have.

In [194]:
# Fill NaN with empty lists to clean hashtags
df_combined['hashtags'] = df_combined['hashtags'].fillna('[]')

# Replace NaN and negative values in num_comments with 0
df_combined.loc[df_combined['num_comments'].isna() | (df_combined['num_comments'] < 0), 'num_comments'] = 0

# Replace NaN and negative values in likes with 0
df_combined.loc[df_combined['likes'].isna() | (df_combined['likes'] < 0), 'likes'] = 0

# For views, only replace NaN and negative values from row 1000 onwards
views = (df_combined.index >= 1000) & (df_combined['views'].isna() | (df_combined['views'] < 0))
df_combined.loc[views, 'views'] = 0

# For followers, only replace NaN and negative values from row 1000 onwards
followers = (df_combined.index >= 1000) & (df_combined['followers'].isna() | (df_combined['followers'] < 0))
df_combined.loc[followers, 'followers'] = 0

In [195]:
problematic_rows = checkForProblematicRows(df_combined)
problematic_rows

,hashtags,num_comments,likes,views,followers


### We will change the type of hashtags to become an actual list for easier processing

In [196]:
import ast

print('Before:', type(df_combined['hashtags'].iloc[1]))

# Convert string representation of list to actual list objects
df_combined['hashtags_list'] = df_combined['hashtags'].apply(ast.literal_eval)

print('After:', type(df_combined['hashtags_list'].iloc[1]))

Before: <class 'str'>
After: <class 'list'>


### Enriching by new attribute has_hashtags

simple classification:
- Empty list: False
- Non-empty lists: True

In [197]:
# Create a category: 'With Hashtags' vs 'Without Hashtags'
df_combined['has_hashtags'] = df_combined['hashtags_list'].apply(lambda x: True if len(x) > 0 else False)

In [198]:
# Drop initial hashtags column because we already have a clean version
df_combined = df_combined.drop(columns=['hashtags'])

df_combined

,num_comments,likes,views,followers,hashtags_list,has_hashtags
0,0,19.0,NaN,NaN,"[#พร้อมส่ง, #สายฝอ, #สายฝอพร้อมส่ง, #สายฝอต้อง...",True
1,14,189.0,NaN,NaN,"[#GanpatiBappaMorya, #GaneshChaturthi, #Festiv...",True
2,0,68.0,NaN,NaN,"[#ramadan, #ramadankareem, #ramadanmubarak, #f...",True
3,1,0.0,NaN,NaN,[],False
4,10,397.0,NaN,NaN,[],False
...,...,...,...,...,...,...
1995,4,493.0,27696.0,31712.0,[],False
1996,0,0.0,339.0,0.0,"[#PINPrestigeMalaysia, #PINMYCover, #covergirl...",True
1997,1,188.0,2381.0,35319.0,"[#turtle, #rehabilitation, #maldives, #reefsca...",True
1998,2,0.0,1115.0,0.0,"[#iOS, #AppleIntelligence, #Apple, #iPhone, #P...",True


### Enriching by new attribute hashtag_count

Get the length of the list to see if the number of hashtags affect the like, comment, view count

In [199]:
# Create the hashtag_count column
df_combined['hashtag_count'] = df_combined['hashtags_list'].apply(lambda x: len(x) if isinstance(x, list) else 0)

df_combined

,num_comments,likes,views,followers,hashtags_list,has_hashtags,hashtag_count
0,0,19.0,NaN,NaN,"[#พร้อมส่ง, #สายฝอ, #สายฝอพร้อมส่ง, #สายฝอต้อง...",True,9
1,14,189.0,NaN,NaN,"[#GanpatiBappaMorya, #GaneshChaturthi, #Festiv...",True,12
2,0,68.0,NaN,NaN,"[#ramadan, #ramadankareem, #ramadanmubarak, #f...",True,18
3,1,0.0,NaN,NaN,[],False,0
4,10,397.0,NaN,NaN,[],False,0
...,...,...,...,...,...,...,...
1995,4,493.0,27696.0,31712.0,[],False,0
1996,0,0.0,339.0,0.0,"[#PINPrestigeMalaysia, #PINMYCover, #covergirl...",True,9
1997,1,188.0,2381.0,35319.0,"[#turtle, #rehabilitation, #maldives, #reefsca...",True,4
1998,2,0.0,1115.0,0.0,"[#iOS, #AppleIntelligence, #Apple, #iPhone, #P...",True,6


## We will need to separate the analysis for posts and reels since there is no data on views and followers, which can potentially skew the results as those big accounts can heavily influence the results we get.

### Analysis to do for all data:
1. Hashtag existence vs engagement:
- Compare comments/likes between posts with hashtags vs without hashtags
- Use t-test or Mann-Whitney U test (since engagement data is likely skewed)

2. Number of hashtags vs engagement:
- Correlation analysis between hashtag count and comments/likes

### Analysis to do for just reels: (1000 - 1999 dataset)
1. Normalize engagement metrics
2. Hashtag existence vs. normalized engagement
3. Number of hashtags vs. normalized engagement

### Extra analysis:
1. Compare the performance of using popular hashtags vs using niche hashtags

### Hashtag existence vs engagement

In [205]:
# Group by hashtag status and look at the average engagement
posts = df_combined.head(1000)
reels = df_combined.tail(1000)

summary_posts = posts.groupby('has_hashtags')[['likes', 'num_comments']].agg(['mean', 'median', 'count'])

summary_reels = reels.groupby('has_hashtags')[['likes', 'num_comments']].agg(['mean', 'median', 'count'])

summary_combined = df_combined.groupby('has_hashtags')[['likes', 'num_comments']].agg(['mean', 'median', 'count'])

print(summary_posts)
print('*****************************************************************')
print(summary_reels)
print('*****************************************************************')
print(summary_combined)

                    likes              num_comments             
                     mean median count         mean median count
has_hashtags                                                    
False         5530.011331   59.0   353    79.926346    2.0   353
True          1925.228748   47.0   647    25.485317    1.0   647
*****************************************************************
                     likes              num_comments             
                      mean median count         mean median count
has_hashtags                                                     
False         15417.281977  207.0   344    96.732558    8.0   344
True           6465.512195  171.0   656    75.524390    7.0   656
*****************************************************************
                     likes              num_comments             
                      mean median count         mean median count
has_hashtags                                                     
False         1

Key Findings:
1. Posts with hashtags are more common
Hashtags present: 1,303 posts (65% of total)
No hashtags: 697 posts (35% of total)
Users clearly prefer using hashtags

2. Counter-intuitive engagement patterns
For Posts (first 1000 rows):
Without hashtags: Higher average likes (5,530 vs 1,925) and comments (79.9 vs 25.5)
But median values tell a different story: median likes (59 vs 47) and comments (2 vs 1) are closer


For Reels (last 1000 rows):
Same pattern: Without hashtags perform better on average
Likes: 15,417 vs 6,465 | Comments: 96.7 vs 75.5

Combined dataset:
Without hashtags: 10,410 average likes vs 4,211 with hashtags
Comments: 88.2 vs 50.7 without hashtags

What This Really Means:
High mean but lower median for non-hashtag posts suggests that very large accounts (celebrities, major brands) often don't need hashtags
These accounts have huge follower bases, so their posts get high engagement regardless of hashtags
This skews the average upward for the "no hashtags" group

For "regular" accounts:
Posts: The same result
Reels: Very close medians (171 with hashtags vs 207 without)
Combined: Almost identical median likes (97 vs 108) and comments (4 vs 4)

Conclusion:
The data suggests that hashtags don't dramatically impact engagement for typical posts, but:
Without hashtags: You get extreme highs (from big accounts) and extreme lows
With hashtags: More consistent, predictable performance
Insight: Hashtags might be more about discoverability than engagement. They help new accounts get found, but once you have a following, they matter less.

In [209]:
from scipy.stats import mannwhitneyu
import numpy as np

def mann_whitney_u_test(df):
    likes_with_hashtags = df[df['has_hashtags'] == True]['likes']
    likes_without_hashtags = df[df['has_hashtags'] == True]['likes']

    # Perform Mann-Whitney U test for likes
    u_stat, p_value_likes = mannwhitneyu(
        likes_with_hashtags, 
        likes_without_hashtags, 
        alternative='two-sided'
    )

    comments_with_hashtags = df[df['has_hashtags'] == True]['num_comments']
    comments_without_hashtags = df[df['has_hashtags'] == False]['num_comments']

    # Perform Mann-Whitney U test for comments
    u_stat_comments, p_value_comments = mannwhitneyu(
        comments_with_hashtags, 
        comments_without_hashtags,
        alternative='two-sided'
    )

    print(f"Likes U-statistic: {u_stat}, p-value: {p_value_likes:.4f}")
    print(f"Comments U-statistic: {u_stat_comments}, p-value: {p_value_comments:.4f}")

In [212]:
print('Posts:')
mann_whitney_u_test(df_combined.head(1000))
print('***********************************************')
print('Reels:')
mann_whitney_u_test(df_combined.tail(1000))
print('***********************************************')
print('Combined:')
mann_whitney_u_test(df_combined)

Posts:
Likes U-statistic: 209304.5, p-value: 1.0000
Comments U-statistic: 109379.5, p-value: 0.2528
***********************************************
Reels:
Likes U-statistic: 215168.0, p-value: 1.0000
Comments U-statistic: 109655.0, p-value: 0.4618
***********************************************
Combined:
Likes U-statistic: 848904.5, p-value: 1.0000
Comments U-statistic: 437031.5, p-value: 0.1590


Hashtags doesnt play much signifcance in boosting engagement

### Number of hashtags vs engagement

In [216]:
# Correlation between hashtag count and engagement
from scipy.stats import spearmanr

def correlation_analysis(df, attribute):
    hashtag_users = df[df['has_hashtags'] == True]
    correlation, p_value = spearmanr(hashtag_users['hashtag_count'], 
                                     hashtag_users[attribute])
    print(f"Correlation between hashtag count and {attribute}: {correlation:.3f}, p={p_value:.3f}")

In [218]:
correlation_analysis(df_combined, 'likes')
correlation_analysis(df_combined, 'num_comments')
correlation_analysis(df_combined.tail(1000), 'views')

Correlation between hashtag count and likes: -0.163, p=0.000
Correlation between hashtag count and num_comments: -0.127, p=0.000
Correlation between hashtag count and views: -0.017, p=0.658


Likes: Weak negative correlation (-0.163, p < 0.001)

Comments: Weak negative correlation (-0.127, p < 0.001)

Views: No correlation (-0.017, p = 0.658)

### Normalise reels data

In [219]:
import pandas as pd
import numpy as np
from scipy.stats import mannwhitneyu, spearmanr
import matplotlib.pyplot as plt
import seaborn as sns

reels_data = df_combined.tail(1000).copy()

# Create normalised metrics (avoiding division by zero)
reels_data['engagement_rate'] = (reels_data['likes'] + reels_data['num_comments']) / reels_data['followers'].replace(0, np.nan) * 100
reels_data['likes_per_view'] = reels_data['likes'] / reels_data['views'].replace(0, np.nan) * 100
reels_data['comments_per_view'] = reels_data['num_comments'] / reels_data['views'].replace(0, np.nan) * 100
reels_data['likes_per_follower'] = reels_data['likes'] / reels_data['followers'].replace(0, np.nan) * 100

# Check the normalised metrics
print("Normalised Metrics Summary:")
print(reels_data[['engagement_rate', 'likes_per_view', 'comments_per_view', 'likes_per_follower']].describe())

Normalised Metrics Summary:
       engagement_rate  likes_per_view  comments_per_view  likes_per_follower
count       519.000000      865.000000         865.000000          519.000000
mean          5.865278       74.788351           3.223845            5.735621
std          39.026506      552.144323          18.883111           38.653903
min           0.000000        0.000000           0.000000            0.000000
25%           0.081852        3.736285           0.018748            0.070186
50%           0.424812        8.343410           0.160772            0.374207
75%           1.526434       18.086957           0.681508            1.486345
max         761.603376    10584.210526         263.636364          756.962025


In [220]:
# Split by hashtag existence
with_hashtags_norm = reels_data[reels_data['has_hashtags'] == True]
without_hashtags_norm = reels_data[reels_data['has_hashtags'] == False]

# Define metrics to test
normalized_metrics = ['engagement_rate', 'likes_per_view', 'comments_per_view', 'likes_per_follower']

# Run Mann-Whitney U tests for each normalized metric
print("=" * 60)
print("MANN-WHITNEY U TEST - HASHTAG EXISTENCE (NORMALIZED DATA)")
print("=" * 60)

results = []
for metric in normalized_metrics:
    # Drop NaN values
    with_data = with_hashtags_norm[metric].dropna()
    without_data = without_hashtags_norm[metric].dropna()
    
    if len(with_data) > 0 and len(without_data) > 0:
        u_stat, p_value = mannwhitneyu(with_data, without_data, alternative='two-sided')
        
        # Calculate medians
        median_with = with_data.median()
        median_without = without_data.median()
        
        results.append({
            'Metric': metric,
            'U-Statistic': u_stat,
            'P-Value': p_value,
            'Significant': p_value < 0.05,
            'Median (With Hashtags)': median_with,
            'Median (Without)': median_without,
            'Difference': median_with - median_without,
            'With N': len(with_data),
            'Without N': len(without_data)
        })

results_df = pd.DataFrame(results)
print(results_df.to_string())

MANN-WHITNEY U TEST - HASHTAG EXISTENCE (NORMALIZED DATA)
               Metric  U-Statistic   P-Value  Significant  Median (With Hashtags)  Median (Without)  Difference  With N  Without N
0     engagement_rate      29636.5  0.518857        False                0.457649          0.374207    0.083442     360        159
1      likes_per_view      80042.5  0.145649        False                7.912415          9.016393   -1.103978     562        303
2   comments_per_view      82530.5  0.454271        False                0.163539          0.149278    0.014260     562        303
3  likes_per_follower      29877.5  0.424794        False                0.402706          0.338780    0.063926     360        159


In [221]:
# Filter to only posts with hashtags and valid normalized metrics
hashtag_users_norm = reels_data[reels_data['has_hashtags'] == True].copy()

print("\n" + "=" * 60)
print("SPEARMAN CORRELATION - HASHTAG COUNT VS NORMALIZED METRICS")
print("=" * 60)

correlation_results = []
for metric in normalized_metrics:
    # Drop NaN values
    valid_data = hashtag_users_norm[[ 'hashtag_count', metric]].dropna()
    
    if len(valid_data) > 0:
        correlation, p_value = spearmanr(valid_data['hashtag_count'], valid_data[metric])
        
        correlation_results.append({
            'Metric': metric,
            'Correlation': correlation,
            'P-Value': p_value,
            'Significant': p_value < 0.05,
            'N': len(valid_data)
        })

correlation_df = pd.DataFrame(correlation_results)
print(correlation_df.to_string())


SPEARMAN CORRELATION - HASHTAG COUNT VS NORMALIZED METRICS
               Metric  Correlation   P-Value  Significant    N
0     engagement_rate     0.018566  0.725532        False  360
1      likes_per_view    -0.035447  0.401631        False  562
2   comments_per_view    -0.013851  0.743179        False  562
3  likes_per_follower     0.013807  0.794038        False  360


### Compare popular vs niche hashtags

In [227]:
import pandas as pd
import numpy as np
from scipy.stats import mannwhitneyu
import matplotlib.pyplot as plt
import seaborn as sns

# First, let's identify the top 20 hashtags from your data
all_hashtags = []
hashtag_lists = df_combined[df_combined['has_hashtags'] == True]['hashtags_list'].tolist()

for hashtag_list in hashtag_lists:
    if isinstance(hashtag_list, list):
        all_hashtags.extend([tag.lower().strip('#') for tag in hashtag_list])

from collections import Counter
hashtag_counts = Counter(all_hashtags)
top_20_hashtags = [tag for tag, count in hashtag_counts.most_common(20)]

print("=" * 70)
print("TOP 20 MOST COMMON HASHTAGS")
print("=" * 70)
for i, tag in enumerate(top_20_hashtags, 1):
    print(f"{i:2d}. #{tag} (used {hashtag_counts[tag]} times)")

print(f"\nTotal unique hashtags: {len(hashtag_counts)}")
print(f"Top 20 represent: {(sum(hashtag_counts[tag] for tag in top_20_hashtags) / len(all_hashtags) * 100):.1f}% of all hashtag uses")

TOP 20 MOST COMMON HASHTAGS
 1. #reels (used 49 times)
 2. #fyp (used 47 times)
 3. #trending (used 44 times)
 4. #photooftheday (used 41 times)
 5. #explore (used 40 times)
 6. #instagood (used 38 times)
 7. #viral (used 38 times)
 8. #instagram (used 38 times)
 9. #love (used 36 times)
10. #fashion (used 33 times)
11. #reelsinstagram (used 32 times)
12. #explorepage (used 28 times)
13. #photography (used 27 times)
14. #trendingreels (used 24 times)
15. #travel (used 22 times)
16. #india (used 21 times)
17. #instadaily (used 20 times)
18. #anime (used 19 times)
19. #art (used 18 times)
20. #style (used 17 times)

Total unique hashtags: 9865
Top 20 represent: 5.1% of all hashtag uses


In [228]:
def contains_top_hashtags(hashtag_list, top_tags):
    """Check if post contains any top 20 hashtag"""
    if not isinstance(hashtag_list, list):
        return False
    
    hashtag_list = [tag.lower().strip('#') for tag in hashtag_list]
    return any(tag in top_tags for tag in hashtag_list)

def count_top_hashtags(hashtag_list, top_tags):
    """Count how many top 20 hashtags are used"""
    if not isinstance(hashtag_list, list):
        return 0
    
    hashtag_list = [tag.lower().strip('#') for tag in hashtag_list]
    return sum(1 for tag in hashtag_list if tag in top_tags)

# Apply to dataset
df_combined['uses_top_hashtag'] = df_combined['hashtags_list'].apply(
    lambda x: contains_top_hashtags(x, top_20_hashtags)
)

df_combined['top_hashtag_count'] = df_combined['hashtags_list'].apply(
    lambda x: count_top_hashtags(x, top_20_hashtags)
)

df_combined['top_hashtag_pct'] = df_combined.apply(
    lambda row: (row['top_hashtag_count'] / len(row['hashtags_list']) * 100) 
    if isinstance(row['hashtags_list'], list) and len(row['hashtags_list']) > 0 else 0, 
    axis=1
)

print("\n" + "=" * 70)
print("POSTS USING TOP 20 HASHTAGS")
print("=" * 70)
print(f"Posts using at least one top 20 hashtag: {df_combined['uses_top_hashtag'].sum()} ({df_combined['uses_top_hashtag'].mean()*100:.1f}%)")
print(f"Posts with NO top 20 hashtags: {(~df_combined['uses_top_hashtag']).sum()} ({(~df_combined['uses_top_hashtag']).mean()*100:.1f}%)")


POSTS USING TOP 20 HASHTAGS
Posts using at least one top 20 hashtag: 261 (13.1%)
Posts with NO top 20 hashtags: 1739 (87.0%)


In [229]:
# Separate datasets
posts_data = df_combined.head(1000).copy()
reels_data = df_combined.tail(1000).copy()

# Calculate normalized metrics for reels
reels_data['engagement_rate'] = (reels_data['likes'] + reels_data['num_comments']) / reels_data['followers'].replace(0, np.nan) * 100
reels_data['likes_per_view'] = reels_data['likes'] / reels_data['views'].replace(0, np.nan) * 100

print("\n" + "=" * 70)
print("PERFORMANCE COMPARISON: TOP HASHTAGS vs OTHERS")
print("=" * 70)

# For Posts (raw numbers)
posts_top = posts_data[posts_data['uses_top_hashtag'] == True]
posts_other = posts_data[posts_data['uses_top_hashtag'] == False]

print("\n--- POSTS (Raw Numbers) ---")
print(f"{'Metric':<20} {'With Top Hashtags':>20} {'Without Top':>20} {'Difference':>15}")
print("-" * 75)

for metric in ['likes', 'num_comments']:
    top_median = posts_top[metric].median()
    other_median = posts_other[metric].median()
    diff_pct = ((top_median - other_median) / other_median * 100) if other_median > 0 else float('inf')
    
    print(f"{metric:<20} {top_median:>20.1f} {other_median:>20.1f} {diff_pct:>+14.1f}%")
    
    # Statistical test
    u_stat, p_value = mannwhitneyu(posts_top[metric], posts_other[metric])
    print(f"{'p-value':<20} {p_value:>20.4f} {'Significant' if p_value < 0.05 else 'Not significant':>20}")

# For Reels (normalized)
reels_top = reels_data[reels_data['uses_top_hashtag'] == True]
reels_other = reels_data[reels_data['uses_top_hashtag'] == False]

print("\n--- REELS (Normalized) ---")
print(f"{'Metric':<20} {'With Top Hashtags':>20} {'Without Top':>20} {'Difference':>15}")
print("-" * 75)

for metric in ['engagement_rate', 'likes_per_view']:
    if metric in reels_top.columns and metric in reels_other.columns:
        top_median = reels_top[metric].median()
        other_median = reels_other[metric].median()
        diff_pct = ((top_median - other_median) / other_median * 100) if other_median > 0 else float('inf')
        
        print(f"{metric:<20} {top_median:>20.2f}% {other_median:>20.2f}% {diff_pct:>+14.1f}%")
        
        # Statistical test
        u_stat, p_value = mannwhitneyu(reels_top[metric].dropna(), reels_other[metric].dropna())
        print(f"{'p-value':<20} {p_value:>20.4f} {'Significant' if p_value < 0.05 else 'Not significant':>20}")


PERFORMANCE COMPARISON: TOP HASHTAGS vs OTHERS

--- POSTS (Raw Numbers) ---
Metric                  With Top Hashtags          Without Top      Difference
---------------------------------------------------------------------------
likes                                47.0                 53.0          -11.3%
p-value                            0.1239      Not significant
num_comments                          1.0                  2.0          -50.0%
p-value                            0.0070          Significant

--- REELS (Normalized) ---
Metric                  With Top Hashtags          Without Top      Difference
---------------------------------------------------------------------------
engagement_rate                      0.42%                 0.42%           -0.0%
p-value                            0.7301      Not significant
likes_per_view                       9.81%                 8.22%          +19.3%
p-value                            0.8844      Not significant
